src/main/java/no/digipost/api/client/Communicator.java

```diff
@@ -72,16 +76,35 @@ public static void checkResponse(final CloseableHttpResponse response, EventLogg
     }
 
     protected static ErrorMessage fetchErrorMessageString(final CloseableHttpResponse response) {
+        final HttpEntity responseEntity = response.getEntity();
+        final StatusLine statusLine = response.getStatusLine();
+        final ErrorType errorType = ErrorType.fromResponseStatus(statusLine);
+        if (responseEntity == null) {
+            return new ErrorMessage(errorType, statusLine + " (respons hadde ikke noe innhold)");
+        }
+
+        final byte[] responseContent;
         try {
-            ErrorMessage errorMessage = unmarshal(errorMessageContext, response.getEntity().getContent(), ErrorMessage.class);
+            responseContent = EntityUtils.toByteArray(responseEntity);
+        } catch (IOException e) {
+            throw new DigipostClientException(ErrorCode.GENERAL_ERROR,
+                    "Server returnerte " + statusLine + ", men klienten feilet ved lesing av responsen: " + e.getMessage(), e);
+        }
+
+        try {
+            ErrorMessage errorMessage = unmarshal(errorMessageContext, new ByteArrayInputStream(responseContent), ErrorMessage.class);
             response.close();
```

Comment:
Is the response closed somewhere else in case of exception?

Reply:
An `AutoCloseable` _given_ as a method argument should _never_ ble closed by the receiving method. It should be handled where (or very near) the `AutoCloseable` is acquired, as this is where it is known when the resources may be disposed.

Maybe the new ability in Java 9 to just say `try (resourceGotFromSomewhereElse) { .. }` will be an enabler for this antipattern 🙄  

I started to refactor for guarantied response closing, and eventually discovered that I also closed the response when there was no error, and the response is going to be further processed outside this method. And would it be ok to close the response if this method discovers that there is an error? Probably, but using this method requires proper resource handling outside of this method anyway.